src/mdns/mdns_avahi.cpp

```diff
@@ -860,6 +919,262 @@ void Publisher::Destroy(Publisher *aPublisher)
+                                                             AvahiProtocol          aProtocol,
+                                                             AvahiBrowserEvent      aEvent,
+                                                             const char *           aName,
+                                                             const char *           aType,
+                                                             const char *           aDomain,
+                                                             AvahiLookupResultFlags aFlags,
+                                                             void *                 aContext)
+{
+    static_cast<PublisherAvahi::ServiceSubscription *>(aContext)->HandleBrowseResult(
+        aServiceBrowser, aInterfaceIndex, aProtocol, aEvent, aName, aType, aDomain, aFlags);
+}
+
+void PublisherAvahi::ServiceSubscription::HandleBrowseResult(AvahiServiceBrowser *  aServiceBrowser,
+                                                             AvahiIfIndex           aInterfaceIndex,
+                                                             AvahiProtocol          aProtocol,
+                                                             AvahiBrowserEvent      aEvent,
+                                                             const char *           aName,
+                                                             const char *           aType,
+                                                             const char *           aDomain,
+                                                             AvahiLookupResultFlags aFlags)
+{
+    OTBR_UNUSED_VARIABLE(aProtocol);
+
+    VerifyOrExit(mServiceBrowser == aServiceBrowser);
+
+    otbrLogInfo("browse service reply: %s.%s%s inf %u, flags=%u", aName, aType, aDomain, aInterfaceIndex, aFlags);
+
+    if (aEvent == AVAHI_BROWSER_FAILURE)
+    {
+        mPublisherAvahi->OnServiceResolveFailed(*this, avahi_client_errno(mPublisherAvahi->mClient));
+    }
+    else
+    {
+        Resolve(aInterfaceIndex, aName, aType, aDomain);
+    }
+exit:
+    if (mServiceBrowser != nullptr)
+    {
+        avahi_service_browser_free(mServiceBrowser);
+        mServiceBrowser = nullptr;
+    }
+}
+
+void PublisherAvahi::ServiceSubscription::Resolve(uint32_t    aInterfaceIndex,
+                                                  const char *aInstanceName,
+                                                  const char *aType,
+                                                  const char *aDomain)
+{
+    otbrLogInfo("resolve service %s %s %s inf %d", aInstanceName, aType, aDomain, aInterfaceIndex);
+    if (!(mServiceResolver = avahi_service_resolver_new(mPublisherAvahi->mClient, aInterfaceIndex, AVAHI_PROTO_INET6,
+                                                        aInstanceName, aType, aDomain, AVAHI_PROTO_INET6,
+                                                        static_cast<AvahiLookupFlags>(0), HandleResolveResult, this)))
```

Comment:
In which thread will `HandleResolveResult` be called?
Can we make sure that `HandleResolveResult` be called in the Thread mainloop?

Reply:
After taking a look at the OTBR code, I think the callbacks are triggered in the OT mainloop.

We have a custom poller to handle all the events, which is single-threaded. The poller is triggered in BorderAgent::Process, which is the mainloop of OT.